### ISSUE TITLE
node.js package

### ISSUE THREAD

**Issue by orlin** [replikativ/replikativ#3]
There are some node.js developers I'd like to inform about replikativ, but for them to consider it, I'd have to say it's on npm.  How would a JS guy use this?  I'm not asking for myself but rather because it says "supposed to work from JavaScript" in the #readme.

I found this [cljs-npm-template](https://github.com/crocket/cljs-npm-template) project recently.  Could it help with making things usable from js (client or server)?  If JS could replicate and work with `CDVCS` it would make things all the more interesting.  Otherwise most people would just look at swarm.js instead.

**Comment by whilo** [comment]
I agree. My plan was first to release it in the clj/cljs community and get some feedback. For contributors it is also necessary atm. to program in Clojure, which makes this community most important. The real potential for applications is with JavaScript (and maybe Java) though. To approach JavaScript there are some steps necessary:
1. expose the APIs in a JavaScript friendly way, that is a) handle all `go-channel` operations via callbacks (which is just syntactic sugar) b) expose `edn` datastructures as JSON or through an edn library. This should then allow to program against replikativ in the browser with Google Closure compiler compatible modules. I am not sure about the status of AMD.js etc. packaging of cljs projects.
2. implement a durable store on `node.js` with konserve, as is done with the [FileStore](https://github.com/replikativ/konserve/blob/master/src/konserve/filestore.clj) on the JVM atm.
3. compile and bundle a `npm` package for node.js developers

Step 1. needs some design work and thinking, therefore I would appreciate some help from a JavaScript person and some feedback before going further. Do you have experience with cljs - js interop?

**Comment by orlin** [comment]
I'd much rather practice clojure[script] right now than battle with node.js interop, which I don't have any experience with anyway.  Also, many other items on the replicativ roadmap are much more important than js use, imo.  From your response / list, it looks more involved than I thought.  I agree the clojure community is what counts atm.  I really enjoyed the clojure group thread and researching this whole topic, which is becoming very important to me, but we could discuss that apart from this issue...

Once js isn't a barrier I bet people would use node.js to replicate data on the server side, a bit more lightweight than jvm.  I guess one could just use cljs for that and deploy to node?  Maybe documentation for such node.cljs backend?  Kind of a backup data server / something to keep things more readily in sync?

**Comment by whilo** [comment]
Sure, feel free to join the gitter chat to discuss :-). I think what is important right now is to demonstrate the power of such a platform. So I discussed with @kordano today that some peer management interface (similar to the CouchDB admin panel) to do time travel and inspection of the peer state might make it more manageable to newcomers. Atm. we have a somewhat fleshed out stage API for the REPL, but in the longer run people will run peers as administrators, e.g. on a raspberry pi with node.js or some JVM backend in the cloud or as a frontend developer on a mobile device. The hottest feature for an admin interface seems time travel to me, which we like to make possible by exposing the commit graph of CDVCS.

The JavaScript API will be important soon, but we also need, configurable subscriptions (atm. everything gets subscribed back), authentication and improved error handling so that peers are resilient. We also would like to implement two CRDTs: a LWWR and grow-sets. I think afterwards tackling JavaScript would not be unreasonable.

**Comment by whilo** [comment]
I have added a first demo client in the `nodejs` folder. There are two issues atm. 1. When replicating the topiq db there is a hashing error, which possibly comes from a string encoding issue and 2. it crashes when sending a committed transaction. This is not ready for packaging yet, but some feedback might already be helpful.
